### DISCUSSION TITLE
EEPROM library flushes data on each write

### DISCUSSION THREAD

**Discussion by me21** [stm32duino/Arduino_Core_STM32#2161] [General]
**Is your feature request/improvement related to a problem? Please describe.**

According to the source code, `EEPROM.write` assigns value to EERef: `(EERef(idx)) = val;`.
This calls `EERef::operator=` which in turn calls `eeprom_write_byte`.
On devices without actual EEPROM, it calls `eeprom_buffered_write_byte` and `eeprom_buffer_flush` which erases and rewrites the whole page of flash leading to excessive wear.

**Describe the solution you'd like**

I like the approach taken in the alternative implementation: https://github.com/rogerclarkmelbourne/Arduino_STM32/tree/master/STM32F1/libraries/EEPROM
Not only it erases the flash page only when needed, it avoids the buffer in RAM by writing half words directly to the free space on the flash page. The new variable value is written to the free space, not overwriting its previous value, thus avoiding the need for flash page erase.
On another hand, that implementation writes a header for each stored variable, halving the available space, whereas this implementation just gives access to the raw contents of the flash page.
Maybe it's possible to add a pair of function like `disableAutoFlush` and `enableAutoFlush`, documenting that each flush causes flash page erase cycle?
Or should I use `eeprom_buffered_write_byte` and `eeprom_buffer_flush` directly? Then `EEPROMClass` is not very much usable with respect to writing data - as I said, each byte change will cause flash page erase.

**Comment by fpistm** [comment]
For me it is covered by https://github.com/stm32duino/Arduino_Core_STM32/pull/333

IIWR Writing to flash one or more bytes required to write full page. 
Do not hesitate to provide a PR if you think having a better solution. 😉

**Comment by me21** [comment]
I see, so your proposal is to use `eeprom_buffered_write_byte(addr, value);` instead of `EEPROM[addr] = value;`.
It's just that `EEPROMClass` currently has not much value as `EEPROM.put(addr, my_struct);` will cause as many page erases as there are changed bytes in `my_struct` since the last write.

**Reply by fpistm** [reply]
Here an example made by a contributor:
https://github.com/stm32duino/STM32Examples/blob/main/examples/NonReg/BufferedEEPROM/BufferedEEPROM.ino
